packages/dds/shared-object-base/src/sharedObject.ts

```diff
@@ -3,450 +3,46 @@
-        // to serialize handles in this object's data. The routes of these serialized handles are outbound routes
-        // to other Fluid objects.
-        assert(!this._isSummarizing, 0x076 /* "Possible re-entrancy! Summary should not already be in progress." */);
-        this._isSummarizing = true;
-
-        let summaryTree: ISummaryTreeWithStats;
-        try {
-            const serializer = new SummarySerializer(this.runtime.channelsRoutingContext);
-            const snapshot: ITree = this.snapshotCore(serializer);
-            summaryTree = convertToSummaryTreeWithStats(snapshot, fullTree);
-            assert(this._isSummarizing, 0x077 /* "Possible re-entrancy! Summary should have been in progress." */);
-        } finally {
-            this._isSummarizing = false;
-        }
-        return summaryTree;
-    }
-
-    /**
-     * {@inheritDoc (ISharedObject:interface).getGCData}
-     */
-    public getGCData(fullGC: boolean = false): IGarbageCollectionData {
-        // Set _isSummarizing to true. This flag is used to ensure that we only use SummarySerializer (created in
-        // getGCDataCore) to serialize handles in this object's data.
-        assert(!this._isSummarizing, 0x078 /* "Possible re-entrancy! Summary should not already be in progress." */);
-        this._isSummarizing = true;
-
-        let gcData: IGarbageCollectionData;
-        try {
-            gcData = this.getGCDataCore();
-            assert(this._isSummarizing, 0x079 /* "Possible re-entrancy! Summary should have been in progress." */);
-        } finally {
-            this._isSummarizing = false;
-        }
-
-        return gcData;
-    }
-
-    /**
-     * Returns the GC data for this shared object. It contains a list of GC nodes that contains references to
-     * other GC nodes.
-     * Derived classes must override this to provide custom list of references to other GC nodes.
-     */
-    protected getGCDataCore(): IGarbageCollectionData {
-        // We run the full summarize logic to get the list of outbound routes from this object. This is a little
-        // expensive but its okay for now. It will be updated to not use full summarize and make it more efficient.
-        // See: https://github.com/microsoft/FluidFramework/issues/4547
-        const serializer = new SummarySerializer(this.runtime.channelsRoutingContext);
-        this.snapshotCore(serializer);
-
-        // The GC data for this shared object contains a single GC node. The outbound routes of this node are the
-        // routes of handles serialized during snapshot.
-        return {
-            gcNodes: { "/": serializer.getSerializedRoutes() },
-        };
+    protected captureStateCore(): ISharedObjectStateHost {
+        const snapshot = this.snapshotCore(this.serializer);
+        return { summarize: async (serializer: IFluidSerializer, fullTree: boolean = false) => {
```

Comment:
I'm a bit lost here RE use of serializer and intentions.
Looking at existing code (which is a bit hard to follow), I believe
- this.serializer was used outside of summaries only. And old code specifically was setting this._isSummarizing = true to ensure that any usage of it in summary would assert.
- summary code was always creating serializer on the fly and was passing it down, but otherwise would destroy it after summary.

I'd think this code needs to do the same - ensure that this.snapshotCore would be using passed in (newly allocated) serializer but would not be able to touch this.serializer. And thus, I'd think summarize() method should not have serializer argument?

I might be wrong, I'm simply trying to follow before & after code and see the differences, intentional or not.

Reply:
The changes are deliberate. We only need to ensure that only the particular, new SummarySerializer is used during GC and never any other serializer. It doesn't matter during non-GC summarization. I verified with Navin.